crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -2,797 +2,535 @@
-                );
-                (digests, None)
-            }
-        };
+            next = &mut stream.next() => {
+                match next {
+                    Some(Ok(BatchInfoResponseItem(UpdateItem::Batch(signed_batch)))) => {
+                        let batch_next_seq = signed_batch.data().next_sequence_number;
+                        debug!(?peer, ?batch_next_seq, "Received signed batch");
+                        metrics.total_batch_received.inc();
+                    }
 
-        // Download the cert and effects - either finality has been establish (above), or
-        // we are a validator.
-        let (cert, effects) = self
-            .download_cert_and_effects(authorities_with_cert, &DownloadRequest::Node(digests))
-            .await?;
-
-        // Node sync request arrive in causal order via the follower API, so it is always safe to
-        // assume that the parents of this cert have already been enqueued.
-        self.wait_for_parents(permit, &digests.transaction, &effects)
-            .await?;
-
-        self.state
-            .handle_node_sync_certificate(cert, effects.clone())
-            .await?;
-
-        // Garbage collect data for this tx.
-        self.effects_stake
-            .lock()
-            .unwrap()
-            .forget_effects(effects.digest());
-        self.node_sync_store.delete_cert_and_effects(digest)?;
-
-        Ok(())
-    }
+                    Some(Ok(BatchInfoResponseItem(UpdateItem::Transaction((seq, digests))))) => {
+                        trace!(?peer, ?digests, ?seq, "received tx from peer");
+                        metrics.total_tx_received.inc();
+                        node_sync_store.enqueue_execution_digests(peer, seq, &digests)?;
+                        process_digest!(seq, digests);
 
-    async fn wait_for_parents(
-        &self,
-        permit: OwnedSemaphorePermit,
-        digest: &TransactionDigest,
-        effects: &SignedTransactionEffects,
-    ) -> SuiResult {
-        // Must drop the permit before waiting to avoid deadlock.
-        std::mem::drop(permit);
-
-        for parent in effects.effects.dependencies.iter() {
-            let (_, mut rx) = self.pending_txes.wait(parent).await;
-
-            if self.state.database.effects_exists(parent)? {
-                continue;
-            }
+                        follow_result.items_from_stream += 1;
+                    }
 
-            trace!(?parent, ?digest, "waiting for parent");
-            // Since we no longer hold the semaphore permit, can be sure that our parent will be
-            // able to start.
-            rx.recv()
-                .await
-                .map_err(|e| SuiError::GenericAuthorityError {
-                    error: format!("{:?}", e),
-                })?;
-        }
+                    Some(Err(err)) => {
+                        debug!(?peer, "handle_batch_stream error: {}", err);
+                        return Err(err);
```

Comment:
We could, but the way this is designed, it shouldn't matter very much, and this keeps the code simpler.

We will restart anything that was in progress, and typically we won't repeat any work. The sequence will be read from the db instead of from the validator, and if the sync process had previously made any progress on other certs (e.g. downloading certs/effects) then nothing will be re-downloaded.